plugins/out_kafka/kafka.c

```diff
@@ -211,6 +213,15 @@ int produce_message(struct flb_time *tm, msgpack_object *map,
             }
         }
 
+        /* Lookup raw_log_key */
+        if (ctx->raw_log_key && !raw_key && val.type == MSGPACK_OBJECT_STR) {
+            if (key.via.str.size == ctx->raw_log_key_len &&
+                    strncmp(key.via.str.ptr, ctx->raw_log_key, ctx->raw_log_key_len) == 0) {
+                raw_key = (char *) val.via.str.ptr;
```

Comment:
this routine is inside a loop, if I am not wrong the values might be set for one message but if the next message does not match they will have the value from the previous one (cc: @cosmo0920 )

Reply:
Ohh, I've overlooked one. Thanks for catching this.

Reply:
How is this different to the `message_key` handling a few lines of above? What is the lifetime of val.via.str.ptr? Shouldn't it be valid until after the msgpack is freed?

Reply:
Normally, we preserve the temporary reference as a heap memory with `flb_sds_create_len`.
In this case, just referencing via `val.via.str.ptr` would be replaced with the latter line in the loop.
So, we need to create a heap region with:


```c
 raw_key = flb_sds_create_len(val.via.str.ptr, val.via.str.size);
```

and in the finalizing function:

```c
flb_sds_destroy(raw_key);
```

Plus, this heap should be used `flb_sds_t` type for it.

Reply:
Took a copy of that memory and freed it on the exit paths. flb_sds_destroy(NULL) seems legit (like free(NULL))